src/collection.rs

```diff
@@ -23,6 +26,143 @@ use strategy::*;
 use tuple::TupleValueTree;
 use test_runner::*;
 
+// TODO: add more CollectionStrategy impls for different types such as
+// HashSet.
+
+// TODO: tests for CollectionStrategy impls.
+
+/// Returns a `Strategy` that generates subsequences of `source` where the
+/// where the subsequences have a size: `size.start <= seq.len() < size.end`.
+pub fn subsequence<T>(source: Vec<T>, size: Range<usize>) -> SubseqStrategy<T>
+where
+    T: Clone + fmt::Debug
+{
+    source.prop_sample(size)
+}
+
+/// A collection, such as `Vec<T>` is itself a strategy to create samples
+/// (or in the case of `Vec`: subsequences) with the elements of the collection.
+pub trait CollectionStrategy: Sized + fmt::Debug {
+    /// The `Strategy` generating the samples.
+    type Strategy: Strategy<Value = Self::ValueTree>;
+
+    /// Always the same as `Self::Strategy::Value`.
+    /// Unfortunately this can't be removed until `impl Trait` lands.
+    type ValueTree: ValueTree<Value = Self>;
+
+    /// Constructs a `Strategy` to generate samples of `self` given a range
+    /// (low..high) where the output sample has a number of elements: len.
+    /// The following always holds for the sample: `low <= len < high`.
+    ///
+    /// # Safety
+    ///
+    /// A panic must occur at least when:
+    ///
+    /// 1. `low > high`,
+    /// 2. `self.len() < low + 1`.
```

Comment:
shouldn't this be `self.len() < low`? I'm not sure if a panic is nessary if `low == len`

Reply:
I was unsure about this too...

but... the problem is how this works:
```rust
        // Generate the amount of elements in result:
        let rng = runner.rng();
        let amount = rand::distributions::Range::new(
            self.size.start, self.size.end).ind_sample(rng);
```

Here `amount \in [self.size.start, self.size.end)`.
Take note of the fact that `self.size.end` is never included in the interval.
So `0..10` will at most give `9`.
Please correct me if I'm wrong on this point - or how (and if) we should solve this.

Reply:
when you say `low` don't you mean `size.start`?

I would say it should panic if:
- `self.len() < size.start`
- `self.len() >= size.end`

I'm getting a little confused about the different names of these.

Reply:
Oh yeah =)

A panic should happen iff: `!(size.start <= self.len() < size.end && size.start < size.end)`.

Reply:
yes, but I would break that up to be more readable :)

Also, doesn't `Range` *already* panic if `size.start > size.end`? Do we really need to specifiy it here?

Reply:
To my knowledge `https://doc.rust-lang.org/nightly/std/ops/struct.Range.html` does not panic when `start > end` and it doesn't hurt to reiterate the requirement.

Reply:
ya, you're right that it doesn't hurt. I actually didn't realize that was a core type in rust.